Refactoring API data pipeline for fetching institutions data and updating Readme

whymath opened the pull request.

whymath committed:
Refactoring fetch institutions from API data pipeline based on PR#16  review

whymath committed:
Updating Readme file for data pipeline refactoring changes

kaaloo commented on collabnext/openalex/institutions.py:
```diff
@@ -1,6 +1,34 @@
 from pyalex import Institution, Institutions
+import json
+import os
+import scripts.fetch_custom_institutions as fetch_custom_institutions
```
Let's leave fetching as a manual step.  This code shouldn't depend on scripts which are more like tools.  There is an open [issue on chaining data loaders](https://github.com/observablehq/framework/issues/332) which if resolved could allow us to move the script functionality to a regular data loader.

whymath replied:
Sounds good, removed API fetch call from `institutions.py` in the new commit

kaaloo commented on collabnext/openalex/institutions.py:
```diff
@@ -1,6 +1,34 @@
+    try:        
+        institutions = json.load(open(institutions_file_path))        
+    except Exception as e:
+        print("\nError loading institutions from JSON file", institutions_file_path, ":", e, "\n")
```
stdout is used for outputting data in an observable framework data loader so it's not possible to use print.  This should work (haven't tried, need to import sys of course):

```suggestion
        print("\nError loading institutions from JSON file", institutions_file_path, ":", e, "\n", file=sys.stderr)
```

whymath replied:
Updated in the new commit

kaaloo commented on collabnext/openalex/institutions.py:
```diff
@@ -1,6 +1,34 @@
+    except Exception as e:
+        print("\nError loading institutions from JSON file", institutions_file_path, ":", e, "\n")
+    
+    # Fetch institutions from API if JSON file is empty or not found
```
Let's just use the random institutions in this case

whymath replied:
Updated in the new commit

kaaloo commented on collabnext/openalex/institutions.py:
```diff
@@ -1,6 +1,34 @@
+    return institutions
+
+
+if __name__ == "__main__":
```
Should'nt need to run this as a module.  Perhaps a unit test would work.  Your call but simply debugging works too.  Here's my vscode launch.json config for debugging the data loader:

```json
    {
      "name": "Debug Data Loader",
      "type": "debugpy",
      "request": "launch",
      "program": "${workspaceFolder}/observable/docs/data/graph.sqlite.py",
      "console": "integratedTerminal",
      "cwd": "${workspaceFolder}/observable"
    }
```

whymath replied:
Thanks, removed this code block in the new commit

kaaloo commented on collabnext/openalex/institutions.py:
```diff
@@ -1,6 +1,34 @@
-def get_institutions() -> list[Institution]:
-    # Get 5 random institutions for now
-    return [Institutions().random() for _ in range(5)]
+def get_institutions(institutions_file_path: str = "observable/docs/data/institutions.json") -> list[Institution]:
```
The path here doesn't work when running `invoke dev` because it runs from within the `observable` directory.

```suggestion
def get_institutions(institutions_file_path: str = "docs/data/institutions.json") -> list[Institution]:
```

whymath replied:
Updated in the new commit

kaaloo reviewed (changes requested):
Fantastic job @whymath !  Thank you so much for all the updates including the improvements to the README.  There are a few issues to attend to but I ran both the script and the data loader.  I'm testing the 85 institutions now, I expect we'll have to select a few when developing because of the time it takes to fetch the data when running `invoke dev`.  As for displaying the data on the graph it could definitely be challenging if it even works.  What would you suggest?  Maybe we can limit to 5 institutions for now.

whymath committed:
Updating data pipeline to limit fetch count and HBCUs priority list

whymath commented:
Makes sense @kaaloo , I have added a new parameter to the script and function so that the number of institutions fetched and loaded is now configurable, with the default value set to 5 in the `invoke fetch` task.

I have also moved the 5 institutions recommended by @lewlefton to the top of the `hbcus_names_list.csv` so they will always be fetched and loaded first. @lewlefton note that I had to update the name of "Alabama A&M University" to "Alabama Agricultural and Mechanical University" since that is how the data is present in OpenAlex.

kaaloo reviewed (approved):
Thank you so much @whymath!  Looks great!

kaaloo commented:
@whymath I'll go ahead and merge your PR since you will only be around a bit later in the day.  Great job!
